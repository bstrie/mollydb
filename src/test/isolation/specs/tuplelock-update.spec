setup {
   DROP TABLE IF EXISTS pktab;
   CREATE TABLE pktab (id int PRIMARY KEY, data SERIAL NOT NULL);
   INSERT INTO pktab VALUES (1, DEFAULT);
}

teardown {
   DROP TABLE pktab;
}

session "s1"
step "s1_advlock" {
    SELECT mdb_advisory_lock(142857),
        mdb_advisory_lock(285714),
        mdb_advisory_lock(571428);
 }
step "s1_chain" { UPDATE pktab SET data = DEFAULT; }
step "s1_begin" { BEGIN; }
step "s1_grablock" { SELECT * FROM pktab FOR KEY SHARE; }
step "s1_advunlock1" { SELECT mdb_advisory_unlock(142857); }
step "s1_advunlock2" { SELECT mdb_sleep(5), mdb_advisory_unlock(285714); }
step "s1_advunlock3" { SELECT mdb_sleep(5), mdb_advisory_unlock(571428); }
step "s1_commit" { COMMIT; }

session "s2"
step "s2_update" { UPDATE pktab SET data = DEFAULT WHERE mdb_advisory_lock_shared(142857) IS NOT NULL; }

session "s3"
step "s3_update" { UPDATE pktab SET data = DEFAULT WHERE mdb_advisory_lock_shared(285714) IS NOT NULL; }

session "s4"
step "s4_update" { UPDATE pktab SET data = DEFAULT WHERE mdb_advisory_lock_shared(571428) IS NOT NULL; }

permutation "s1_advlock" "s2_update" "s3_update" "s4_update" "s1_chain" "s1_begin" "s1_grablock" "s1_advunlock1" "s1_advunlock2" "s1_advunlock3" "s1_commit"
