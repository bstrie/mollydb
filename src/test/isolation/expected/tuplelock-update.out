Parsed test spec with 4 sessions

starting permutation: s1_advlock s2_update s3_update s4_update s1_chain s1_begin s1_grablock s1_advunlock1 s1_advunlock2 s1_advunlock3 s1_commit
step s1_advlock: 
    SELECT mdb_advisory_lock(142857),
        mdb_advisory_lock(285714),
        mdb_advisory_lock(571428);

mdb_advisory_lockmdb_advisory_lockmdb_advisory_lock

                                             
step s2_update: UPDATE pktab SET data = DEFAULT WHERE mdb_advisory_lock_shared(142857) IS NOT NULL; <waiting ...>
step s3_update: UPDATE pktab SET data = DEFAULT WHERE mdb_advisory_lock_shared(285714) IS NOT NULL; <waiting ...>
step s4_update: UPDATE pktab SET data = DEFAULT WHERE mdb_advisory_lock_shared(571428) IS NOT NULL; <waiting ...>
step s1_chain: UPDATE pktab SET data = DEFAULT;
step s1_begin: BEGIN;
step s1_grablock: SELECT * FROM pktab FOR KEY SHARE;
id             data           

1              2              
step s1_advunlock1: SELECT mdb_advisory_unlock(142857);
mdb_advisory_unlock

t              
step s2_update: <... completed>
step s1_advunlock2: SELECT mdb_sleep(5), mdb_advisory_unlock(285714);
mdb_sleep       mdb_advisory_unlock

               t              
step s3_update: <... completed>
step s1_advunlock3: SELECT mdb_sleep(5), mdb_advisory_unlock(571428);
mdb_sleep       mdb_advisory_unlock

               t              
step s4_update: <... completed>
step s1_commit: COMMIT;
